install.hs

```diff
@@ -167,15 +164,13 @@ configureCabal versionNumber = do
     ["new-configure", "-w", ghcPath, "--write-ghc-environment-files=never"]
 
 findInstalledGhcs :: IO [(VersionNumber, GhcPath)]
-findInstalledGhcs = foldM
-  (\found version -> do
-    path <- getGhcPath version
-    case path of
-      Nothing -> return found
-      Just p  -> return $ (version, p) : found
+findInstalledGhcs = mapMaybeM
+  (\version -> do
+    getGhcPath version >>= \case
+      Nothing -> return Nothing
+      Just p  -> return $ Just (version, p)
   )
-  []
-  hieVersions
+  (reverse hieVersions)
```

Comment:
What about something along the lines of this untested code:
```hs
findInstalledGhcs = reverse <$> mapMaybeM getBothPathAndVersion hieVersions where
    getBothPathAndVersion version = addVersion version <$> getGhcPath version
    addVersion version maybePath = (version,) <$> maybePath
```

It's somewhat convoluted that the two `<$>` are for different functors, but I think "maybe" in `maybePath` clarifies that.

You  can also use `catMaybes` from Data.Maybe from `base` to avoid the dependency

Reply:
> What about something along the lines of this untested code:
> 
> ```haskell
> findInstalledGhcs = reverse <$> mapMaybeM getBothPathAndVersion hieVersions where
>     getBothPathAndVersion version = addVersion version <$> getGhcPath version
>     addVersion version maybePath = (version,) <$> maybePath
> ```
> 
> It's somewhat convoluted that the two `<$>` are for different functors, but I think "maybe" in `maybePath` clarifies that.
> 
> You can also use `catMaybes` from `Data.Maybe` from `base` to avoid the dependency

Imo this is kind of hard to read. I would prefer readability over conciseness.